File: app/src/main/java/com/v2/desafionubank/adapter/RecyclerViewAdapter.java
package com.v2.desafionubank.adapter;


import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;


/**
 * Created by CaioSChristino on 07/02/2018.
 *//**/

public abstract class RecyclerViewAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private Context mContext;
    private List<T> mItems = new ArrayList<>();

    public RecyclerViewAdapter(Context mContext) {
        this.mContext = mContext;
    }

    public RecyclerViewAdapter(Context mContext, List<T> mItems) {
        this.mContext = mContext;
        this.mItems = mItems;
    }

    public abstract RecyclerView.ViewHolder setViewHolder(ViewGroup parent);

    public abstract void onBindData(RecyclerView.ViewHolder holder, T item);

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder holder = setViewHolder(parent);
        return holder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        onBindData(holder, mItems.get(position));
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    public void addItems(List<T> items) {
        this.mItems = items;
        this.notifyDataSetChanged();
    }

    public void addItems(T item) {
        this.mItems.add(item);
        this.notifyDataSetChanged();
    }

    public Context getmContext() {
        return mContext;
    }

    public List<T> getmItems() {
        return mItems;
    }
}

File: app/src/main/java/com/v2/desafionubank/adapter/RecyclerViewAdapter.kt
package com.v2.desafionubank.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import java.util.*

/**
 * Created by CaioSChristino on 18/04/18.
 */

abstract class RecyclerViewAdapter<T> : RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private var mContext: Context? = null
    private var mItems: MutableList<T> = ArrayList()

    constructor(mContext: Context) {
        this.mContext = mContext
    }

    constructor(mContext: Context, mItems: MutableList<T>) {
        this.mContext = mContext
        this.mItems = mItems
    }

    abstract fun setViewHolder(parent: ViewGroup): RecyclerView.ViewHolder

    abstract fun onBindData(holder: RecyclerView.ViewHolder, item: T)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return setViewHolder(parent)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        onBindData(holder, mItems[position])
    }

    override fun getItemCount(): Int {
        return mItems.size
    }

    fun addItems(items: MutableList<T>) {
        this.mItems = items
        this.notifyDataSetChanged()
    }

    fun addItems(item: T) {
        this.mItems.add(item)
        this.notifyDataSetChanged()
    }

    fun getmContext(): Context? {
        return mContext
    }

    fun getmItems(): List<T> {
        return mItems
    }
}